addon/components/frost-popover.js

```diff
@@ -1,65 +1,96 @@
-      this.set('visibility', false)
-      $('html').off('click.container.' + this.get('elementId'))
-      if (_.isFunction(action)) {
-        action()
+    close () {
+      if (this.get('isDestroyed')) { return }
+      this.set('visible', false)
+    },
+    togglePopover () {
+      this.toggleProperty('visible')
+      let position = this.get('position')
+      let targetRect
+      // eslint-disable-next-line max-len
+      let targetElement = this.get('closest') ? this.$().closest(this.get('target'))[this.get('index')] : this.get('parentView').$(this.get('target'))[this.get('index')]
+      targetRect = {
+        top: targetElement.offsetTop,
+        left: targetElement.offsetLeft,
+        bottom: targetElement.offsetTop + targetElement.offsetHeight,
+        right: targetElement.offsetLeft + targetElement.offsetWidth,
+        width: targetElement.offsetWidth,
+        height: targetElement.offsetHeight
+      }
+      let popoverElement = this.get('element')
+      let popoverRect = popoverElement.getBoundingClientRect()
+      let top
+      let left
+      let excludePadding = this.get('excludePadding')
+      switch (position) {
+        case 'bottom':
+          top = targetRect.bottom + this.get('offset')
+          left = targetRect.left + targetRect.width / 2 - popoverRect.width / 2
+          if (excludePadding) {
+            let cs = window.getComputedStyle(targetElement)
+            top -= parseInt(cs.getPropertyValue('padding-bottom'))
+          }
+          popoverElement.style.top = top + 'px'
+          popoverElement.style.left = left + 'px'
+          break
+        case 'top':
+          top = targetRect.top - popoverRect.height - this.get('offset')
+          left = targetRect.left + targetRect.width / 2 - popoverRect.width / 2
+          if (excludePadding) {
+            let cs = window.getComputedStyle(targetElement)
+            top += parseInt(cs.getPropertyValue('padding-top'))
+          }
+          popoverElement.style.top = top + 'px'
+          popoverElement.style.left = left + 'px'
+          break
+        case 'left':
+          top = targetRect.top + targetRect.height / 2 - popoverRect.height / 2
+          left = targetRect.left - popoverRect.width - this.get('offset')
+          if (excludePadding) {
+            let cs = window.getComputedStyle(targetElement)
+            left += parseInt(cs.getPropertyValue('padding-left'))
+          }
+          popoverElement.style.top = top + 'px'
+          popoverElement.style.left = left + 'px'
+          break
+        case 'right':
+          top = targetRect.top + targetRect.height / 2 - popoverRect.height / 2
+          left = targetRect.right + this.get('offset')
+          if (excludePadding) {
+            let cs = window.getComputedStyle(targetElement)
+            left -= parseInt(cs.getPropertyValue('padding-right'))
+          }
+          popoverElement.style.top = top + 'px'
+          popoverElement.style.left = left + 'px'
```

Comment:
Since you're doing the same thing in all cases, it might be better to have this statement after the switch case.